File: _102_BinaryTreeLevelOrderTraversal.java
// Time Complexity : o(n) where n is number of nodes in tree
// Space Complexity : o(n) where n is number of nodes in tree
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// Approach : Iterative lvel order traversal using queue, add the root to queue and while q is not empty get the size of queue and
//remove the elements , add the children of removed elements back in queue
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {

        List<List<Integer>> res = new ArrayList<>();

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while(!q.isEmpty()){

            int size = q.size();
            List<Integer> temp = new ArrayList<>();

            for(int i = 0; i<size; i++){
                TreeNode curr  = q.poll();
                if(curr != null){
                    temp.add(curr.val);
                    q.add(curr.left);
                    q.add(curr.right);
                }
            }

            if(temp.size() > 0){
                res.add(temp);
            }
        }

        return res;
    }
}

File: 102_BinaryTreeLevelOrdertraversal.cpp
// Time Complexity : O(N)
// Space Complexity :O(N)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(root == NULL){
            vector<vector<int>> x;
            return x;
        }
        vector<vector<int>> res;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int size = q.size();
            vector<int> li;
            for(int i=0; i<size; i++){
                TreeNode* curr = q.front();
                q.pop();
                li.push_back(curr->val);
                if(curr->left != NULL)
                    q.push(curr->left);
                if(curr->right != NULL)
                    q.push(curr->right);
            }
            res.push_back(li);
        }
        return res;
    }
};


// Time Complexity : O(N)
// Space Complexity :O(H) where H is the height of the tree.
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> result;
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(root == NULL){
            return result;
        }
        dfs(root, 0);
        return result;
    }
    void dfs(TreeNode* root, int level){
        if(root==NULL)
            return;
        
        if(level == result.size()){
            result.push_back({root->val});
        }
        else{
            result[level].push_back(root->val);
        }
        dfs(root->left, level+1);
        dfs(root->right, level+1);
    }
};
